crates/datasources/src/flightsql/mod.rs

```diff
@@ -0,0 +1,152 @@
+mod insert;
+mod scan;
+
+use std::any::Any;
+use std::sync::Arc;
+
+use arrow_flight::sql::client::FlightSqlServiceClient;
+use arrow_flight::sql::CommandGetDbSchemas;
+use async_trait::async_trait;
+use datafusion::arrow::datatypes::SchemaRef;
+use datafusion::datasource::TableProvider;
+use datafusion::error::DataFusionError;
+use datafusion::execution::context::SessionState;
+use datafusion::logical_expr::{Expr, TableProviderFilterPushDown, TableType};
+use datafusion::physical_plan::ExecutionPlan;
+use datafusion_ext::errors::ExtensionError;
+use tonic::transport::{Channel, Endpoint};
+
+use crate::common::query;
+use crate::common::util::Datasource;
+
+// use arrow_flight::FlightClient;
+
+// use arrow_flight::error::FlightError;
+// use arrow_flight::encode::FlightDataEncoderBuilder;
+// use datafusion::physical_plan::execute_stream;
+// use futures::StreamExt;
+
+#[derive(Debug, Clone)]
+pub struct FlightSqlSourceConnectionOptions {
+    pub uri: String,
+    pub database: String,
+    pub table: String,
+    pub token: String,
+}
+
+pub struct FlightSqlSourceProvider {
+    channel: Channel,
+    schema: SchemaRef,
+    opts: FlightSqlSourceConnectionOptions,
+}
+
+impl FlightSqlSourceProvider {
+    pub async fn try_new(opts: FlightSqlSourceConnectionOptions) -> Result<Self, ExtensionError> {
+        let uri = opts.uri.clone();
+        let channel = Endpoint::from_shared(uri)
+            .map_err(|e| ExtensionError::String(e.to_string()))?
+            .connect()
+            .await
+            .map_err(|e| ExtensionError::String(e.to_string()))?;
+
+
+        let db = opts.database.clone();
+        let mut client = Self::make_client(opts.clone(), channel.clone());
+        let catalogs = client.get_catalogs().await?;
+        let schema = Arc::new(
+            client
+                .get_db_schemas(CommandGetDbSchemas {
+                    catalog: Some(catalogs.to_string()),
+                    db_schema_filter_pattern: Some(db),
+                })
+                .await?
+                .try_decode_schema()?,
+        );
```

Comment:
Does this return the correct schema? This doesn't seem right to me.

Reply:
yeah, probably should be a table thing.